pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java

```diff
@@ -801,4 +810,12 @@ public void setupOutput(ContextImpl contextImpl) throws Exception {
             Thread.currentThread().setContextClassLoader(this.instanceClassLoader);
         }
     }
+
+    public String getStatsAsString() throws IOException {
```

Comment:
We should synchronize this method as well?  

Also line 241 (run() method) accesses the stats variable and the close() method does too. Difference threads access these methods and change the stats variable. It's the same for other variables in this class.

Reply:
yes. Done

Reply:
There is a follow up change coming up that makes stats class threadsafe.

Reply:
Actually I've made the stats class synchronized to ensure that it becomes threadsafe.
Also have moved the setup syncronized, so that it and close won't be able to run together